codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -480,12 +480,14 @@ private void cancel(Throwable cause) {
                 return;
             }
 
+            // If cause == null this method was called via onStreamClosed(...) or onStreamHalfClosed(...)
+            Http2Error error = cause == null ? Http2Error.STREAM_CLOSED : INTERNAL_ERROR;
```

Comment:
`STREAM_CLOSED` has a specific meaning defined by the RFC and its use here may not follow the RFC's definition.

https://tools.ietf.org/html/rfc7540#section-7
> The endpoint received a frame after a stream was half-closed.

Reply:
Using the STREAM_CLOSED seems fine to me. If we sent these frames, then the remote endpoint _would_ have "received a frame after the stream was half-closed." But I'm not as wild about the implicit error choice based on the cause. Could we pass that in explicitly instead? Functionally that will be the same.

However, that does expose one problem: when there is half-close and still-pending frames, that should kill the channel.

Reply:
Sure I can also pass it in directly if it makes stuff more clear.

So you say if it’s half closed we should close it ?

Reply:
If it's half closed and there are messages pending, then it seems good to close the channel.

Reply:
@Scottmitch agree ?

Reply:
I guess yeah we should also take AUTO_CLOSE into account here. Thats a good idea. Let me do this

Reply:
@Scottmitch so I had a deeper look and I think we can detect the difference here. When we receive a RST_FRAME the `Http2FrameListener.onRstStreamRead(...)` is called. So I think it should be ok ?